test-tests/shared/src/test/scala/zio/test/GenSpec.scala

```diff
@@ -443,6 +443,42 @@ object GenSpec extends ZIOBaseSpec {
     },
     testM("runHead") {
       assertM(Gen.int(-10, 10).runHead)(isSome(isWithin(-10, 10)))
+    },
+    testM("crossAll") {
+      val gen = Gen.crossAll(
+        List(
+          Gen.fromIterable(List(1, 2)),
+          Gen.fromIterable(List(3)),
+          Gen.fromIterable(List(4, 5))
+        )
+      )
+      assertM(gen.runCollect)(
+        equalTo(
+          List(
+            List(1, 3, 4),
+            List(1, 3, 5),
+            List(2, 3, 4),
+            List(2, 3, 5)
+          )
+        )
+      )
+    },
+    testM("zipAll") {
+      val gen = Gen.zipAll(
+        List(
+          Gen.fromIterable(List(1, 2)),
+          Gen.fromIterable(List(3)),
+          Gen.fromIterable(List(4, 5))
+        )
+      )
+      assertM(gen.runCollect)(
+        equalTo(
+          List(
+            List(1, 3, 4),
+            List(2, 3, 5)
```

Comment:
I'm surprised this one is not `List(2, 5)` because I assumed the `List(3)` would be exhausted. Does `Gen.fromIterable` create an infinite `Gen` that cycles the given iterable?

Reply:
`zipWith` repeats pulling from the shorter stream if necessary until the longer stream is exhausted.  In this specific case if we're combining multiple generators we could just return a shorter list, but in general if we are combining `Gen[A]` and `Gen[B]` to produce `Gen[(A, B)]` we can't do that because we need both an `A` and a `B`.

This is important because we model random generators as generators of a single effectual value, so if we combine a random generator with a deterministic generator we want to return a new generator with the size of the original deterministic generator that pulls each value from the deterministic generator, pulling a value from the random generator each time.

For combining deterministic generators both approaches would work I think but this also ends up being fairly useful because you make sure that you sample the entire space of the generators you are combining.

I tried to make this clear in the documentation, for example for `zipAll`, but let me know if there is something you would change.

```
  /**
   * Zips the specified generators together pairwise. The new generator will
   * generate elements as long as any generator is generating elements, running
   * the other generators multiple times if necessary.
   */
```

Reply:
Ah interesting. That reasoning makes sense to me 👍 . Thanks for expanding on that.